### ISSUE TITLE
Enhancement request - partial edges for low cost production boards and prototyping?

### ISSUE THREAD

**Issue by splitn2** [mikemag/MonkeyCAM#17]
For rapid prototyping and as a customer choice for a lower cost production board I would like to have an option to select the program to cut the base for partial edges just along the full running length so a minimal amount of bending just into the tip and tail transition area where the bezier meets the arc.  Worth it or a PITA? cheers Rich

**Comment by splitn2** [comment]
@Mike I think I have another keen coder, he might lean on your for some C tutorship but he's keen and a very passionate mountaineer and skier and I suspect he's better at coding than he lets on! He likes the idea of lightweight edges ie. short edges less steel. I'll get him in touch ok mate. cheers

**Comment by RegMeasures** [comment]
Hi @mikemag I've started playing with this. I'm keen and have a fair bit of programming experience but have no experience with git and little with C++ so this will be a learning process. I figure this shouldn't be too bad an issue to start on. 

As I see it what needs doing is:

1. Add parameters to loadBoard function (in [MonkeyCam.cpp](https://github.com/mikemag/MonkeyCAM/blob/master/cli/monkey-cam.cpp)) to describe the length/position of the partial edge. I'm thinking along the lines of two parameters, one for the distance from the nose to the start of the metal edge and a second for the distance from tail.

2. Modify the generateBaseCutout function in [shapes.cpp](https://github.com/mikemag/MonkeyCAM/blob/master/core/shapes.cpp) to limit the region where the base shape is offset inwards to the length where we want the metal edge to be. To do this I guess I'll have to clip the overall board shape and the current base shape (offset in by 2mm) using some simple rectangles, then recombine them. I can work this out by following what you've already done to generate the core shape where it meets at the nose and tail spacers as this is relatively similar.

Does that approach sound reasonable? Do you have any better ideas for how to define the edge length and position?

Cheers, Reg

**Comment by splitn2** [comment]
Hey @RegMeasures awesome to have you involved mate, I'm sure Mike will give you excellent guidance when necessary inbetween the other parts of his very busy life.

For the partial edges, from a constructors point of view what I am keen on is to have no additional bending of edges if possible, so just letting the edge fall in to the sidecut without then at each end having to go in the other direction into the tip and tail at all, otherwise we start having to bend edges again, and I hate bending edges. So I would think keep the length just within the actual finished sidecut radius....

@mikemag has done this in MonkeyCAM ver 3 too?

Managed to cut some stuff badly today but was a good start, just getting a feel for what I don't know and a huge learning curve ahead!!!!!

**Comment by mikemag** [comment]
@RegMeasures, glad you're excited to help!!! Some quick answers to your questions:

Your suggestion for how to form a new base with partial edges is reasonable. In the past (v3) I put in the effort to ensure that I cut the end of each edge section perpendicular to the edge at that point, but it's overkill. We're using round cutters, so we have to clean that part of the base up with a knife anyway. So clipping with a simple square should suffice.

To decide where to end the edge, measure from the end of the effective, not from the end of the nose/tail. No one knows how long the arcs of their noses and tails are :) What you'll want to do is say "extend the edge 4cm past the effective edge". You may want to adjust that based on the nose or tail geometry, so a different setting for nose and tail would be good. By doing this people will end up with consistency between different board shapes: when you stand up all your boards after a year the edges will end a similar distance from the "hips". It's most likely that you'll pick a default length, and most people will never have to change it. 

Make sure to make a boolean option for "use partial edges" vs "full wrap". 

@splitn2 suggests that he wants to end his edges right at the end of the effective edge. So make sure that 0 is a valid value for the new parameters. That said, it's probably a bad idea to end right there. You want to go past the hips a bit, otherwise you'll have the rough, vulnerable edge end right at the leading end of the effective edge, which is engaged with the snow surface in a carve. @splitn2, going a few cm past the ends of the EE will be minimal edge bending effort, don't worry!! :)

**Comment by mikemag** [comment]
Oh, and @RegMeasures, be sure to see the sections on https://github.com/mikemag/MonkeyCAM/wiki on modifying MonkeyCAM if you haven't already. There's a good overview of the overall architecture and how shapes are formed.

**Comment by splitn2** [comment]
@mikemag @RegMeasures thats great guys !!!!!!

**Comment by splitn2** [comment]
@RegMeasures I have some paulownia .....how is your coding going? I'll start ripping timber this weekend and building up core blocks, we can try this short edge on your skis once you have your ski design in MonkeyCAM. the cnc is ready and I'll have the timber and other materials too. cheers, Rich

**Comment by RegMeasures** [comment]
Nice work @splitn2. I've got the partial edges working - need to send it back to @mikemag to merge it in. I would have done this earlier but I got bogged down trying to fix a bug with the line endings in the G-Code output files which I think is related to compiling on windows. I can't seem to work out why it is doing it though...

**Comment by splitn2** [comment]
cool @RegMeasures I look forward to seeing this. As soon as we have a new release and then with your ski design you have your first set of MonkeyCAM output files, I suggest we do a trial cut and see how the files behave.

**Comment by splitn2** [comment]
@RegMeasures last night I did a test run of my machine , just "cutting air", of a rifle stock design which is another collaboration project I'm working on, and the results were pleasing, the CNC seems to be very responsive and powerful. We were using a basic CAM out of Solidworks which is only good for evaluation but it helped prove the point I'm making the right progress. I have a load of cutting tools coming from china and my vacuum dust extraction system is almost dialed. So progress in the workshop is good. Keep up your good work mate this is exciting!

**Comment by mikemag** [comment]
Fixed by @RegMeasures with ac3eede63d7607c7ab938683b04c7f5142bf3113.

**Comment by mikemag** [comment]
Included in v4.0.4, https://github.com/mikemag/MonkeyCAM/releases/tag/4.0.4
